### ISSUE TITLE
Inline JSDoc parameter type info for arrow-functions is mistaken for JSDoc documentation on the function

### ISSUE THREAD

**Issue by WhyEvenTry** [eslint/eslint#5231]
eslint 2.0

Because WebStorm is able to use the type information I use inline JSDoc type info (which is valid JSDoc). Such type-info of course is just that - it has nothing to do with documenting a function (@param,@returns) - it's just type information.

Second, I use inline arrow functions and I leave out the parenthesis when there is only a single parameter to such an inline arrow function.

Example:

```
// Without () around the single parameter: eslint problem!
Promise.resolve(42)
.then(/**number*/ n => n * 2);

// With () around the parameter "n"
Promise.resolve(42)
.then((/**number*/ n) => n * 2);
```

Problem: In the forst example eslint assumes that I'm trying to document a function and reports 

     error  Missing JSDoc @returns for function                      valid-jsdoc

![Screenshot](https://cloud.githubusercontent.com/assets/199179/13026894/7df7da4e-d239-11e5-99d7-6c722d39d50e.png)

**Comment by ilyavolodin** [comment]
I'm not all that familiar with jsDoc but it does look like this shouldn't be an issue. @gyandeeps should this be parsed as jsDoc at all?

**Comment by WhyEvenTry** [comment]
Just an aside, I'm not sure that syntax comes from JSDoc - I think it comes from the Google Closure Compiler (there are examples for these inline-types on https://developers.google.com/closure/compiler/docs/js-for-compiler). Anyway, WebStorm makes no distinction, it supports and calls "JSDoc support" everything that is in either the official JSDoc documentation or on the Google Closure Compiler type syntax page (just linked).

**Comment by nzakas** [comment]
It will parse as just a description, not as a type. I'm not sure this is something we can fix because there's no good way for us to tell if something is a one-word description or a type.

**Comment by WhyEvenTry** [comment]
I don't understand. It sees that type - and suddenly wants me to document the inline(!) function with a full JSDoc comment. I don't expect or need it to parse as a "type".

**Comment by ilyavolodin** [comment]
@WhyEvenTry point is, we see jsDoc comment, the rule specifies that we should require `return` for all valid function jsDoc comments. In this case, we have nothing to tell us that this isn't really a valid jsDoc comment and we should just ignore it. Basically we can't differentiate between those two cases:
```js
/** Helper */
function a() {
}
```
and
```js
(/**number*/ n) => n * 2
```

**Comment by WhyEvenTry** [comment]
But your example is not what my code/my example says.

**Comment by nzakas** [comment]
@WhyEvenTry these are just examples. What he's saying is that we do not have a way to know if a JSDoc block is meant as an inline type or not. People can put complete JSDoc comments where you have the inline type, and there's no good way for us to know what you intended.

**Comment by WhyEvenTry** [comment]
If you can see that a function is *inline* you can prevent the "return required" rule - which makes no sense in any setup in that situation? I have yet to see anyone trying to fully document - with @param and @return(s) - an inline function. That's not in any example for either JSDoc or Google Closure Compiler either. Inline you just document types.

If you think keeping the mechanism to require @return(s) inside inline functions is what you want, how about a switch instead of just 0,1,2 in eslintrc to disable its enforcement in inline functions?

**Comment by WhyEvenTry** [comment]
I experimented and slept over it a bit more. I have `require-jsdoc` turned OFF - so any errors are only due to rule `valid-jsdoc`.

Two observations:

1) A simple type-only JSDoc coment *is* valid JSDoc.

2) I have in my `valid-jsdoc` rule

    "requireReturn": false,

yet I get errors like

    166:7   error  Missing JSDoc @returns for function                      valid-jsdoc


So as I see it, the rule is called "valid JSDoc". My JSDoc *is* valid. It should not complain. I have *not* enabled the options that "require return".

**Comment by WhyEvenTry** [comment]
Regarding the reporting of `Missing JSDoc @returns for function` despite having the rule option `"requireReturn": false`, is the if statement in line 280 (https://github.com/eslint/eslint/blob/master/lib/rules/valid-jsdoc.js#L280) correct, should that be && instead of ||?

                if (requireReturn || functionData.returnPresent) {
                    context.report(jsdocNode, "Missing JSDoc @" + (prefer.returns || "returns") + " for function.");
                }

I switched the conditoon to an && and now it works: I get errors when I require return statements and I don't get errors when I don't.

**Comment by ilyavolodin** [comment]
@WhyEvenTry `requireReturn` is poorly named options. I ran into the same issue recently as well, and it should be named something like `requireReturnOnlyIfNecessary`, as in, `requireReturn:true` will always require return, but `false` will only require it if the function has a return statement (in your case, you are using arrow functions, which have implicit return).
We also use a different parser for jsDoc, and can't support all of the Google Closure jsDoc enhancements (or WebStorm's as it seems like it follows more of a Google Closure standards).

**Comment by nzakas** [comment]
This is a bit confusing, so I think it's worth documenting this edge case around arrow functions for the rule.
